File: cpp/09_StringToInteger.cpp
/*
 * Q9. String to Integer (atoi) (Category: Strings)
 *
 * Problem: Implement myAtoi — convert a string to a 32-bit signed integer.
 *          Handle leading whitespace, optional sign, non-digit termination, overflow.
 *
 * Time Complexity : O(n)
 * Space Complexity: O(1)
 */
#include <iostream>
#include <string>
#include <climits>
using namespace std;

int myAtoi(string s) {
    int i = 0, n = s.length();
    long result = 0;
    int sign = 1;

    // Skip leading whitespace
    while (i < n && s[i] == ' ') i++;

    // Handle optional sign
    if (i < n && (s[i] == '+' || s[i] == '-')) {
        sign = (s[i] == '-') ? -1 : 1;
        i++;
    }

    // Read digits and clamp overflow
    while (i < n && isdigit(s[i])) {
        result = result * 10 + (s[i] - '0');
        if (result * sign >= INT_MAX) return INT_MAX;
        if (result * sign <= INT_MIN) return INT_MIN;
        i++;
    }
    return (int)(result * sign);
}

int main() {
    cout << myAtoi("42")             << endl; // 42
    cout << myAtoi("   -42")         << endl; // -42
    cout << myAtoi("4193 with words")<< endl; // 4193
    cout << myAtoi("-91283472332")   << endl; // -2147483648
    return 0;
}

File: python/09_string_to_integer.py
"""
Q9. String to Integer (atoi) (Category: Strings)

Problem: Implement myAtoi — convert a string to a 32-bit signed integer.
         Handle leading whitespace, optional sign, non-digit termination, overflow.

Time Complexity : O(n)
Space Complexity: O(1)
"""

INT_MAX = 2**31 - 1   #  2147483647
INT_MIN = -(2**31)    # -2147483648


def my_atoi(s: str) -> int:
    i, n = 0, len(s)
    result = 0
    sign = 1

    # Skip leading whitespace
    while i < n and s[i] == " ":
        i += 1

    # Handle optional sign
    if i < n and s[i] in ("+", "-"):
        sign = -1 if s[i] == "-" else 1
        i += 1

    # Read digits
    while i < n and s[i].isdigit():
        result = result * 10 + int(s[i])
        if result * sign >= INT_MAX:
            return INT_MAX
        if result * sign <= INT_MIN:
            return INT_MIN
        i += 1

    return result * sign


if __name__ == "__main__":
    print(my_atoi("42"))              # 42
    print(my_atoi("   -42"))          # -42
    print(my_atoi("4193 with words")) # 4193
    print(my_atoi("-91283472332"))    # -2147483648
